### ISSUE TITLE
Call for beta-testing reviewers

### ISSUE THREAD

**Issue by timokau** [timokau/marvin-mk2#34]
I just pushed a new version with a new [usage](https://github.com/timokau/marvin-mk2/blob/master/USAGE.md) model. This new model now makes it possible to assign reviewers and automate triage. For now I am the [only reviewer](https://github.com/timokau/marvin-mk2/blob/04b9494479e8f6ff9632683bff52d77199fd82e5/marvin/team.py#L89). I'd like to have a handful more to scale up the test. Not too many though, as everything is still in development.

As a reviewer you can set a limit on how many active nixpkgs PRs you want to be involved in. For example if you have already created 2 PRs today and reviewed one more, you are involved in 3 PRs in the period of one day. If you have set a limit of 4PRs/day, marvin will still request one additional review.

4PRs is likely too much, start slow. Be aware that you're not only signing up for reviews, but also for marvin-mk2 beta testing. Things might go wrong, though I try to avoid that of course.

For now, I will get pinged on every PR that is marked as `needs_merger`. Your reviews will lead to something. Once that gets too much for me, I'll look for help on that front too.

The easiest way to get started is to just make a PR and add yourself [here](https://github.com/timokau/marvin-mk2/blob/04b9494479e8f6ff9632683bff52d77199fd82e5/marvin/team.py#L89). The format may not be the most understandable, so I can also do that for you if you prefer.

Any interest @cole-h, @turion, @doronbehar, @glittershark, @symphorien, @fgaz, @evils, @JJJollyjim, @bbigras, @mdlayher?

**Comment by evils** [comment]
i was under the impression the intended reviewer interface was looking up the label [like this](https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+is%3Apr+label%3Aneeds_reviewer)
What is the effect of being a listed reviewer?

**Comment by timokau** [comment]
Yes, its both. Being listed as a reviewer will actively request your reviews on `needs_reviewer` PRs. It will only do so if you don't already have a lot of activity (determined [like this](https://github.com/NixOS/nixpkgs/issues?q=involves%3Aevils+updated%3A%3E%3D2020-07-01+-merged%3A%3C2020-07-01)) in nixpkgs recently, respecting everybody's "open source time and energy budget".

That is supposed to establish a stable "baseline throughput" of reviews. I want to shift away a bit from the burst-review model to something more sustainable, like a lot of people each reviewing a PR every day or every couple of days. Actively assigning reviews seems more realistic than expecting people to actively look for a single PR to review every x days.

If the capacity of the registered reviewers is not sufficient (which it probably won't be), that can be complemented by going through the `needs_reviewer` list actively when you have some additional time.

**Comment by doronbehar** [comment]
> The easiest way to get started is to just make a PR and add yourself here. The format may not be the most understandable, so I can also do that for you if you prefer.
> Any interest

You can add me as a reviewer. I think I won't be available to start out as soon as possible. Just to be clear, are we talking about actual Nixpkgs or the playground?

**Comment by timokau** [comment]
> You can add me as a reviewer.

What rate-limit would you like? As described above, the bot will count how many active (within `d` days) PRs you are currently involved in and only request reviews from you when that number is below a limit `l`. So you need to set a `d` and a `l`. For example, you are [involved](https://github.com/NixOS/nixpkgs/issues?q=involves%3Adoronbehar+updated%3A%3E%3D2020-07-04+-merged%3A%3C2020-07-04) in 5 PRs with `d=1`.

> I think I won't be available to start out as soon as possible.

So you don't want to be added? ;)

> Just to be clear, are we talking about actual Nixpkgs or the playground?

Nixpkgs proper. The playground is only as a first test for basic functionality before deploying. Its a bit too contrived for an actual practical test, only baptism by fire will do here.

**Comment by fgaz** [comment]
Feel free to add me as well, with a more long-term d=5 l=7

**Comment by glittershark** [comment]
Perhaps might be a good idea to limit reviewers based on the attributes of PRs, too? For example some reviewers (eg me) have hardware running darwin but not others.

Regardless, happy to review one or two PRs / week, however that translates to the `d` and `l` parameters

**Comment by turion** [comment]
I'd like to participate with d=7 and l=3. This is exciting!

**Comment by symphorien** [comment]
I'd also like to participate with d=7 and l=3

**Comment by doronbehar** [comment]
> So you don't want to be added? ;)

Yea @timokau I'm a bit internally conflicted as my semester is about to finish and my tests are up coming. I'll ping you in 3 weeks or so when I'll feel ready to fully dive in.

Besides that, I really like:

> As described above, the bot will count how many active (within d days) PRs you are currently involved in and only request reviews from you when that number is below a limit l

Which is damn sophisticated. :1st_place_medal:  :)

**Comment by timokau** [comment]
> Perhaps might be a good idea to limit reviewers based on the attributes of PRs

Maybe, at some point in the future. I'm not entirely sure its desirable though. Usually you won't need a darwin-specific reviewer, only maybe if the darwin build is failing.

Consider yourself a PR shepherd when marvin requests a "review" from you. Your goal is to push the PR forward in some way, that may mean reviewing it yourself or delegating to someone else if there is some specific requirement.

> Regardless, happy to review one or two PRs / week, however that translates to the `d` and `l` parameters

That leaves some room for interpretation. It could mean that you want two review requests from marvin (not possible to set that since marvin doesn't remember who it requested where) or that you want to be "involved" in two PRs that are not your own or that you want to be "involved" in two PRs period.

I've chosen the most conservative [third](https://github.com/NixOS/nixpkgs/issues?q=involves%3Aglittershark+updated%3A%3E%3D2020-06-30+-merged%3A%3C2020-06-30) interpretation for now in #51.

**Comment by timokau** [comment]
We now have 6 reviewers, great!

> I'm a bit internally conflicted as my semester is about to finish and my tests are up coming.

I understand, I'm in a similar situation. No pressure.

> Which is damn sophisticated. :1st_place_medal: :)

Thanks :) I'm not sure if its scalable enough, since it requires firing of a GH search before requesting anybody as a reviewer. GH ratelimits those pretty tightly. Some caching will probably help there though once we're running into issues.

**Comment by turion** [comment]
> > Perhaps might be a good idea to limit reviewers based on the attributes of PRs
> 
> Maybe, at some point in the future. I'm not entirely sure its desirable though. Usually you won't need a darwin-specific reviewer, only maybe if the darwin build is failing.

I agree. If the first reviewer thinks there is some special Darwin (or other platform) knowledge necessary, then this is not the first step in the review.

**Comment by turion** [comment]
Is there any mechanism to find a different "PR shepherd" if the first one isn't fit for the job (e.g. inappropriate responses of some kind)?

**Comment by timokau** [comment]
> Is there any mechanism to find a different "PR shepherd" if the first one isn't fit for the job (e.g. inappropriate responses of some kind)?

You can manually set the status back to `needs_reviewer`. Marvin will also do this automatically when the reviewer is unresponsive (in `awaiting_reviewer` or `awaiting_merger` for 3 days without any update).

Setting the status to `needs_reviewer` will trigger a triage run, which will assign a new reviewer if available. Older PRs have precedence though.

**Comment by timokau** [comment]
In case anybody was missing any bot actions: A bug crashed the triage runner. I forgot that you can only request reviews from people who have been "explicitly granted read permissions" to the repo, i.e. people in the maintainer list. Marvin attempted to request a review from someone who wasn't a maintainer. I have implemented a fallback for that case and re-deployed.

**Comment by timokau** [comment]
By the way, we are currently at capacity. So if a PR isn't immediately being assigned a reviewer, that is probably not a bug:

```
Jul 09 16:31:38 marvin-mk2: Running triage early
Jul 09 16:31:39 marvin-mk2: Running triage on NixOS/nixpkgs
Jul 09 16:31:39 marvin-mk2: Timing out awaiting_reviewer PRs
Jul 09 16:31:39 marvin-mk2: Timing out awaiting_merger PRs
Jul 09 16:31:40 marvin-mk2: Assigning mergers to needs_merger PRs
Jul 09 16:31:40 marvin-mk2: Assigning reviewers to needs_reviewer PRs
Jul 09 16:31:40 marvin-mk2: Selecting reviewer from candidates: ['ryantm', 'glittershark', 'symphorien', 'timokau', 'fgaz', 'turion']
Jul 09 16:31:40 marvin-mk2: Testing symphorien
Jul 09 16:31:41 marvin-mk2: Active in 18 PRs, limit is 3.
Jul 09 16:31:41 marvin-mk2: Testing fgaz
Jul 09 16:31:41 marvin-mk2: Active in 7 PRs, limit is 7.
Jul 09 16:31:41 marvin-mk2: Testing ryantm
Jul 09 16:31:42 marvin-mk2: Active in 3 PRs, limit is 1.
Jul 09 16:31:42 marvin-mk2: Testing turion
Jul 09 16:31:42 marvin-mk2: Active in 4 PRs, limit is 3.
Jul 09 16:31:42 marvin-mk2: Testing timokau
Jul 09 16:31:42 marvin-mk2: Active in 5 PRs, limit is 1.
Jul 09 16:31:42 marvin-mk2: Testing glittershark
Jul 09 16:31:43 marvin-mk2: Active in 5 PRs, limit is 2.
Jul 09 16:31:43 marvin-mk2: No reviewer found for #92336.
```

**Comment by glittershark** [comment]
Does it check authored PRs? I'm not currently active in reviewing any PRs, only PRs I wrote

**Comment by timokau** [comment]
Yes, it counts everything that github calls "involves" (authored/left a comment/has an outstanding review  request/approved/...). You don't have to be active yourself in the PR either, you just have to be "involved" at some point and the PR has to be active. The assumption is that you're probably still following it and are getting notifications about it.

**Comment by turion** [comment]
> Yes, it counts everything that github calls "involves" (authored/left a comment/has an outstanding review request/approved/...). You don't have to be active yourself in the PR either, you just have to be "involved" at some point and the PR has to be active. The assumption is that you're probably still following it and are getting notifications about it.

Just to be sure: But only those PRs count that have had activity in the last d days?

**Comment by timokau** [comment]
Right, that is what I mean by active. To be precise, anything github counts as an "update" in the last d days. Activity after merge doesn't count (since otherwise it would count as an activity when the PR author decides to delete the branch after merge).

The criteria could be amended of course (even on a per-reviewer basis), as long as the info is easily obtainable by GitHub. The intention behind the current criteria is to have some approximation of your general "involvement" in nixpkgs and not to overload your plate to the point where reviews feel like too much of an annoying chore. Intuitively, after having received 100 emails from PRs I am subscribed to and having responded to 20 reviews on my own PRs, I don't want to be asked to review 5 more ;)

**Comment by timokau** [comment]
Marvin briefly ignored the rate limit due to a bug. Fix is deployed (#80). Sorry for the inconvenience.

**Comment by turion** [comment]
I haven't been notified by the bot for a long time now. Is it working properly?

**Comment by fgaz** [comment]
Same, but it may be because I'm already over the limit I set just from my normal activity

**Comment by turion** [comment]
It would be cool if I could inspect what the bot currently perceives as my activity, so I can increase my limit to a reasonable number.

**Comment by timokau** [comment]
Yes, everybody's limit is exceeded most of the time. @turion you have set it to 3 activities in 7 days, so that can easily be saturated by just day-to-day activity. You can use the [search term](https://github.com/NixOS/nixpkgs/issues?q=involves%3Aturion+updated%3A%3E%3D2020-07-01+-merged%3A%3C2020-07-01) I have posted a couple of times already to get an estimate. In the last 7 days, you had activity in 15 MRs. That is more than 3 ;)

**Comment by timokau** [comment]
Also just to [re-post](https://github.com/timokau/marvin-mk2/issues/86#issuecomment-688514598) it here: Marvin has been on my backburner for a while now, which is why there hasn't been much progress. It has happily chugged along though, no issues.

**Comment by timokau** [comment]
[Re](https://github.com/timokau/marvin-mk2/pull/92#issuecomment-714367067)-posting a current excerpt of the logs here, which might be interesting:

> Oct 22 09:16:28 marvin-mk2: Selecting reviewer from candidates: ['symphorien', 'timokau', 'fgaz', 'glittershark', 'kevincox', 'turion']
> Oct 22 09:16:28 marvin-mk2: Skipping pr author turion
> Oct 22 09:16:28 marvin-mk2: Testing timokau
> Oct 22 09:16:32 marvin-mk2: Limit (1/1d) exceeded until 2020-10-22 19:45:14+00:00.
> Oct 22 09:16:32 marvin-mk2: Testing fgaz
> Oct 22 09:16:32 marvin-mk2: New event: #95530 issue_comment->created
> Oct 22 09:16:35 marvin-mk2: Limit (2/7d) exceeded until 2020-10-28 16:32:36+00:00.
> Oct 22 09:16:35 marvin-mk2: Testing symphorien
> Oct 22 09:16:38 marvin-mk2: Limit (3/7d) exceeded until 2020-10-27 11:47:09+00:00.
> Oct 22 09:16:38 marvin-mk2: Testing kevincox
> Oct 22 09:16:42 marvin-mk2: Limit (2/7d) exceeded until 2020-10-27 20:52:47+00:00.
> Oct 22 09:16:42 marvin-mk2: Testing glittershark
> Oct 22 09:16:45 marvin-mk2: Limit (2/7d) exceeded until 2020-10-28 19:30:56+00:00.
> Oct 22 09:16:45 marvin-mk2: No reviewer found for #95165.

I have come to think that the current rate-limiting approach probably tries to be too clever. I plan to transition to assigning people to PRs (instead of requesting reviews). This would then make it possible to query assigned & open PRs for reviewers and set a limit that way. I think that would be more useful and easier to understand.

This will most likely take a while though, I don't spend much time on marvin recently. Its also somewhat blocked by https://github.com/NixOS/rfc39/issues/3, although it would be possible to work around that.

**Comment by kevincox** [comment]
I agree that rate-limiting is probably not the best approach. I think the max-in-flight approach you described makes sense. Although it also isn't perfect as if I leave a comment and the author doesn't respond for a while then I am probably fine taking another issue. I guess this could be solved with some sort of inactivity timeout?

I'm not too sure about assignee. The fact that I left a comment or two isn't a solid indication that it is the authors turn again. For example I might say that I need to check what other packages do or I want to ask in the discourse or similar. Probably just asking people to submit the review with a status (not just leaving a comment) would be good. However IIRC "request changes" can make it annoying if someone else wants to step in and finish your review (maybe you went on vacation for a week). So maybe that approach isn't perfect either 🤔

Maybe if the bot just handled the author -> review assignee transition it would work? Then the reviewer can manually assign back? (Or submitting a "request changes" review can do it automatically?)

**Comment by timokau** [comment]
> Although it also isn't perfect as if I leave a comment and the author doesn't respond for a while then I am probably fine taking another issue.

Yes, I was thinking to only consider PRs with recent activity.

> I'm not too sure about assignee. The fact that I left a comment or two isn't a solid indication that it is the authors turn again. For example I might say that I need to check what other packages do or I want to ask in the discourse or similar. Probably just asking people to submit the review with a status (not just leaving a comment) would be good. However IIRC "request changes" can make it annoying if someone else wants to step in and finish your review (maybe you went on vacation for a week). So maybe that approach isn't perfect either thinking

I don't understand what exactly you are commenting on. The assignee vs review requested thing is independent from the state transitions. The state transitions are just a heuristic, its not very important when they are inaccurate. It only matters if the PR becomes inactive, in which case marvin nags on a awaiting_reviewer PR but does nothing on an awaiting_changes PR. In that case the status can still be corrected manually. The heuristic errs on the side of awaiting_reviewer for that reason (to avoid having PRs starved on reviews).

Assigning ist mostly convenient because its easier to query for assignees than reviewers and because it matches the intention more clearly. The idea is that the assignee is a "PR shepherd", who can review the PR if they are able to but can also delegate the review if that seems better.

> Maybe if the bot just handled the author -> review assignee transition it would work? Then the reviewer can manually assign back? (Or submitting a "request changes" review can do it automatically?)

I do not understand, could you expand on that?

**Comment by kevincox** [comment]
> The assignee vs review requested thing is independent from the state transitions.

Ah I see, I interpreted this as you using the assignee to decide what the state was and how many reviews are active.

> The idea is that the assignee is a "PR shepherd"

If you look at it that way it makes sense. However GitHub has dashboards that show things you are assigning to GitHub-wide. So it will probably break some workflows if people find them assigned to things that aren't actionable. Reviewers are much better understood by GitHub to be maybe-actionable depending on the state. So it may be less against-the-grain to use reviewers for this.

> I do not understand, could you expand on that?

I meant that if the bot is going to move the assignee we might consider making the `review -> assignee` transition explicit by the reviewer. However I now see that this wasn't the plan.

**Comment by timokau** [comment]
> > The assignee vs review requested thing is independent from the state transitions.
> 
> Ah I see, I interpreted this as you using the assignee to decide what the state was and how many reviews are active.

No, just as a replacement for "review requested". It would be used to see how many reviews are active though, simply by querying "how many PRs that are assigned to this person have been updated in the last 5 days" or some such. That is not currently possible with the "review requested".

> > The idea is that the assignee is a "PR shepherd"
> 
> If you look at it that way it makes sense. However GitHub has dashboards that show things you are assigning to GitHub-wide. So it will probably break some workflows if people find them assigned to things that aren't actionable. Reviewers are much better understood by GitHub to be maybe-actionable depending on the state. So it may be less against-the-grain to use reviewers for this.

That is a very interesting datapoint for consideration, thank you for that! I still feel like the advantages outweigh the disadvantages though. I think it would be just as "actionable" as any other PR you might be assigned to. It might be different for issues, but PRs just aren't always actionable for reviewers.

> > I do not understand, could you expand on that?
> 
> I meant that if the bot is going to move the assignee we might consider making the `review -> assignee` transition explicit by the reviewer. However I now see that this wasn't the plan.

Ah, no that would be the same thing.

**Comment by timokau** [comment]
Marvins docs were a bit lacking and distributed all over the place. I have rewritten the README to fix that. Its already merged, but feedback is still welcome: https://github.com/timokau/marvin-mk2/pull/95

**Comment by kisik21** [comment]
Please add me too, I'd like for some simple PRs going my way to acquire some code review experience. `d=2 l=4` for now - this'll be a low limit so the bot wouldn't annoy me if I'm not active or reviewed some things already, but will still ping me from time to time if I can't find things to do here.

**Comment by timokau** [comment]
Welcome :tada: (https://github.com/timokau/marvin-mk2/pull/98)

**Comment by Ekleog** [comment]
I just came across this issue, and the feature described here looks awesome.

I'd like to volunteer for beta-testing (with d=3 l=1 for reviews and d=3 l=3 for merges, starting slow to test my ability to actually respond), but have a few things I'd like to check before, as I don't trust myself and would rather not end up blocking reviews:
- If I don't notice marvin's mention or am unavailable for a while, will marvin eventually give the review to someone else?
- If I tag as changes requested, will marvin re-ping me (or another reviewer) when the status switches again to needs_review, or am I expected to remember the PRs I'm involved with to re-review them once the comments are handled?

Regardless of the answers to these questions I'm willing to volunteer with the above parameters, but depending on the answers I may end up blocking review-able PRs, so I'm not sure it'd be the best option.

Overall, I'm feeling like we'd benefit greatly from speaking more about marvin and encouraging people to register for reviews, but maybe I'm underestimating marvin's current rough edges :)

**Comment by timokau** [comment]
> I just came across this issue, and the feature described here looks awesome.

:)

> I'd like to volunteer for beta-testing (with d=3 l=1 for reviews and d=3 l=3 for merges, starting slow to test my ability to actually respond), but have a few things I'd like to check before, as I don't trust myself and would rather not end up blocking reviews:
> 
>     * If I don't notice marvin's mention or am unavailable for a while, will marvin eventually give the review to someone else?

Marvin will post a reminder after a PR has been in "awainting reviewer" for 3 days without any changes. After 3 more days it will automatically put the PR back in the "needs reviewer" queue and assign a new reviewer if available. Or at least it should, the second part is actually broken right now (https://github.com/timokau/marvin-mk2/issues/99).

>     * If I tag as changes requested, will marvin re-ping me (or another reviewer) when the status switches again to needs_review, or am I expected to remember the PRs I'm involved with to re-review them once the comments are handled?

It will not immediately re-ping you and it will not ping another reviewer. My assumption is that you will get email notifications for the changes. It will however remind you after three days (same functionality as above).

> Regardless of the answers to these questions I'm willing to volunteer with the above parameters, but depending on the answers I may end up blocking review-able PRs, so I'm not sure it'd be the best option.

Great, welcome! (https://github.com/timokau/marvin-mk2/pull/101)

> Overall, I'm feeling like we'd benefit greatly from speaking more about marvin and encouraging people to register for reviews, but maybe I'm underestimating marvin's current rough edges :)

I agree, but I'm biased ;) There are a few rough edges however:

- The main one is the incredibly unintuitive rate limiting (see https://github.com/timokau/marvin-mk2/issues/34#issuecomment-714386315).
- Then there's the broken timeout. That one should be a small fix, but I haven't had the time to deal with the "test, deploy and verify nothing explodes in production" overhead yet.
- Which brings me to my third priority (which is less of a user-facing rough edge): Make third party contributions possible with reasonable effort. This whole thing is probably not going to work out if I am the only maintainer. Part of that plan would be
	- Restructure the architecture into something more testable: A logical core, a command sink and an event source. That probably only works as neatly in my imagination though. Sounds suspiciously like something that might never happen, even to me
	- Move the team declaration into the target repository (nixpkgs). That makes it easier to test on other repositories and removes me as the bottleneck for changes to the rate limit.
	- Create a guide on setting up a marvin instance on heroku with a test repository (like https://github.com/timokau/marvins-playground/).

All in all, I just have too little time to work on it. Feel free to promote it while keeping these caveats in mind :)

**Comment by lovesegfault** [comment]
Could you add me to the team as well? I can review and merge :)

**Comment by timokau** [comment]
Nice! What rate-limit do you want for each role? Marvin will run a search query like this one:

> [involves:lovesegfault updated:>=2020-12-19T11:00:00 -merged:<2020-12-19T11:00:00 sort:updated-desc](https://github.com/NixOS/nixpkgs/issues?q=involves%3Alovesegfault+updated%3A%3E%3D2020-12-19T11%3A00%3A00+-merged%3A%3C2020-12-19T11%3A00%3A00+sort%3Aupdated-desc+)

Which lists all nixpkgs discussions that you are involved in and that were active in a given time frame (7 days in this case, since 2020-12-19T11:00:00 was 7 days ago).

It will count the results in this query and only request a review / merge-review if it is below a threshold. Right now there are 17 results, so any "limit" greater than that will likely result in a request once a week.

You need to pick a limit  (something like `limit=20, days=7`) for reviews and another one for merge-reviews. Yes, I know that is way more complicated than it needs to be :smile:

**Comment by lovesegfault** [comment]
* Merge: `limit=7, days=7`
* Review: `limit=1, days=3` 

@timokau :)

**Comment by timokau** [comment]
Great, done (#105)!

**Comment by thiagokokada** [comment]
I want to be added to this beta. Sadly I don't have merge permissions, but I can do the reviews at least.

Let's try `days=7 limit=5`. I have a good amount of PRs reviewed at the end of year so it will be probably a no-op until sometime later.

**Comment by timokau** [comment]
Done: https://github.com/timokau/marvin-mk2/pull/109 :rocket:

**Comment by SuperSandro2000** [comment]
I am just going to add myself in vault #111.

**Comment by kisik21** [comment]
can I temporarily step down from the beta until further notice? some circumstances in my life (namely, burnout) prevent me from interacting with Nixpkgs as much as I would like, and reviewing has temporarily become too much for me.

**Comment by timokau** [comment]
Of course! See https://github.com/timokau/marvin-mk2/pull/115. Thank you. I hope you'll be able to take some time for yourself.

**Comment by kevincox** [comment]
Hi, commenting from the issue just referenced. I was the original requested reviewer for that PR but unfortunately I didn't feel adequately able to review the PR. I think there are two main problems here.

1. While I can review most PRs in nixpkgs there are a couple of corners where I don't feel comfortable. I think we need to allow reviews to step out.
2. Your comment seems to suggest that the bot tried to "re-assign" me, which wasn't great in the first place but also gave me no indication of that.

My first thought for how to address this would be something like the following:

1. Try to assign to maintainers first.
  a. There are other bots that do this so maybe we can rely on them?
  b. Current maintainer detection IIUC is by nixpkgs metadata, which doesn't work great for modules, tests and similar. Maybe we could have a fallback mechanism such as CODEOWNERS files that was more reliable for critical parts of the tree (like the Linux kernel).
2. Have an explicit command for a reviewer to opt out, whether they are unable to review due to knowledge or temporary unavailability.

I hope we can keep the bot going. I think it does a lot of good overall but I understand that it can be hard to run it as a 1-man show. (And unfortunately I don't have the ability to help out right now.)

**Comment by turion** [comment]
>     1. While I can review most PRs in nixpkgs there are a couple of corners where I don't feel comfortable. I think we need to allow reviews to step out.

I asked that question some time ago on discourse, and the answer was, paraphrased:

It's fine not to be knowledgeable in all parts, you can still be a good reviewer. 

* If you come across something unfamiliar, you can always try to summon an additional interviewer who is expert on the topic. You might find them by looking at maintainers of similar files, git history, community activity and so on. They can answer a specific question, while it's still your job to drive the process. In that situation, you're more like the assignee of a MR (i.e. the person responsible for progress) than the reviewer (i.e. the person responsible for a good review). My take: Sometimes this works great, sometimes noone answers.
* Take PRs with unfamiliar content as a chance to learn something new. Take your time, familiarise yourself with everything. Possibly use the previous steps to figure out parts which you're still unsure about. My take: Works great in theory, and sometimes also in practice, but not if you don't have a lot of time and energy at hand. (But if you don't have time and energy, you should probably cut down on volunteering anyways somewhere.)

Another thought for having better matches between PRs and assigned reviewers would be opt-in and opt-out of specific labels. E.g. I might want to review only PRs with the haskell label in them, or never PRs with the python label.

**Comment by timokau** [comment]
For those who only follow by email notifications, here is the relevant part of my comment [1] that started the discussion:

> It seems like the bot is not really sustainable with me as the sole maintainer given my current priorities. Maybe it would be better to shut it down. It would be sad to abort this experiment without exploring it a bit further, but that may still be better than the current state. The bot is too spammy in its current configuration and may actually put off contributors or generate unnecessary stress. I think I will post a comment to [timokau/marvin-mk2#34](https://github.com/timokau/marvin-mk2/issues/34) soon-ish and ask for peoples opinions.

[1] https://github.com/NixOS/nixpkgs/pull/121518#issuecomment-898009229

**Comment by kevincox** [comment]
On 13/08/2021 02.58, Manuel Bärenz wrote:
>
> It's fine not to be knowledgeable in all parts, you can still be a 
> good reviewer.
>
I agree. And I definitely plan on following this review. But I was just 
way over my head. Not only was the code complex (that's fine I can deal 
with it) but there appears to be some domain-knowledge changes. I don't 
think I can meaningfully help with those with my current knowledge. I 
did suggest a reviewer that touched this code recently so that we can 
get an expert opinion and I could learn.

While I agree that you can still be helpful in these cases I think there 
is a limit. In this case my limit was low enough that I needed a second 
opinion.

I don't think that the goal is to encourage "punt on hard reviews" I 
just think it is important to have a graceful way to request another 
opinion. Whether that is due to lack of required domain knowledge or 
temporary unavailability.

**Comment by timokau** [comment]
>     1. While I can review most PRs in nixpkgs ohere are a couple of corners where I don't feel comfortable. I think we need to allow reviews to step out.

I agree with @turion's suggestions, but I also agree that volunteers need much more control over the review requests. That's part of why I am thinking of shutting the bot down: The current way it handles things is likely a source of stress for reviewers.

Reviewers should be able to

- opt out of individual reviews with *very low friction*. They currently kind of can opt-out, but that doesn't work perfectly (e.g. the bot may re-request the review) and isn't low friction. This needs both technical and non-technical improvements. On the non-technical side we need a social standard that emphasizes it's okay to opt-out any time, without any need to explain yourself. That could be encouraged with proper wording and configuration of the bot.
- configure which PRs they want to review as @turion suggested. There are advantages to everybody being pinged on everything (it spreads knowledge around, non-expert reviewers may actually be in a better position to require good docs), but people may still prefer some kinds of PRs over others. The bot should respect individual preferences.
- configure how many PRs they want to review. This is currently possible, but has *way* to much friction. It requires a PR to marvin-mk2 and a deployment. Configuration through nixpkgs (as part of the maintainer list?) would be better, but still not ideal. People are unlikely to do that in high-stress phases, and that may be an additional source of stress. At some point I had a prototype that allowed decentralized configuration through a personal github gist. Something like that would be better.

>     2. Your comment seems to suggest that the bot tried to "re-assign" me, which wasn't great in the first place but also gave me no indication of that.

Yes, that is #86.

> My first thought for how to address this would be something like the following:
> 
>     1. Try to assign to maintainers first.
>        a. There are other bots that do this so maybe we can rely on them?
>        b. Current maintainer detection IIUC is by nixpkgs metadata, which doesn't work great for modules, tests and similar. Maybe we could have a fallback mechanism such as CODEOWNERS files that was more reliable for critical parts of the tree (like the Linux kernel).

My intention was that the marvin-assigned reviewer is more like a "shepherd" who can then pull-in additional reviewers (like the maintainers) where it makes sense. As I already mentioned, having everyone review everything has some benefits. But I agree that assigning maintainers first where possible is probably a better use of our limited people-power.

>     2. Have an explicit command for a reviewer to opt out, whether they are unable to review due to knowledge or temporary unavailability.

Yes, definitely.

> I hope we can keep the bot going. I think it does a lot of good overall but I understand that it can be hard to run it as a 1-man show. (And unfortunately I don't have the ability to help out right now.)

There are a lot of low-hanging fruit for improvements. The lowest-hanging one would be to finally fix the automatic reset to "needs\_reviewer". We should also decrease the nagging frequency, make it easier to contribute (#113) and improve automatic testing.

The problem isn't that there is no room for improvement. It is that it has been a few months now, and I haven't gotten around to making these improvements yet. I feel bad about these long-unfixed rough edges, but at the same time it's unlikely to become a high enough priority for me any time soon.

I already argued that the bot may stress out reviewers. Leaving these issues unfixed stresses me out as well. Then there's the risk of something going wrong with the bot. It should really be maintained more actively, but I'm not in a position to do that right now. That's why I think it may be better to shut it down.

If anyone wants to give it a try, contributions would be very welcome. Unfortunately getting up and running is not trivial right now. #113 would be a good starting point.

**Comment by timokau** [comment]
I agree @kevincox. As I wrote in my last comment, I think reviewers should have much more flexibility on how they choose to volunteer their time.

(Replying separately because we were apparently writing our comments at the same time.)

**Comment by LunNova** [comment]
I would like to help maintain this, will look into implementing improvements which improve friction for reviewers.

**Comment by timokau** [comment]
That would be great! Please make sure you test your changes on a playground repository first though, see #113. Verification is the currently the greatest hurdle when making changes. Making that part easier would probably be the change with the highest leverage and would also be a requirement for leaving the experimental stage at some point.

**Comment by timokau** [comment]
I have just disabled marvin-mk2 by setting the github webhook to inactive.

Marvin was an interesting experiment, but it is not without its issues right now. It tends to spam PRs, does not have an easy mechanism to temporarily opt-out in stressful times and in general feels a bit rough around the edges. See [this post](https://github.com/timokau/marvin-mk2/issues/34#issuecomment-898558265) for details on its current shortcomings.

It was never intended to exist in this state for as long as it has. It was supposed to be a proof of concept, to be iterated on and eventually discussed by the wider community. Unfortunately I do not have the bandwidth to properly maintain it right now, let alone improve it.

My thanks to all the "beta-testing reviewers" and everybody else who has helped along the way. We have certainly helped some PRs move along faster than they otherwise would have. Thank you for volunteering your time!

**Comment by turion** [comment]
Thank you @timokau for spending a lot of time and effort into this experiment! I think it was very helpful to understand better what kind of infrastructure `nixpkgs` might profit from.
